ADBDEV-4726-35 Check that scalarVal is not null

dnskvlnk opened the pull request:
Check that scalarVal is not null

appendValue and appendKey dereference scalarVal without checking that it's not
null. We can't guarantee that scalarVal is always valid. This patch adds Insist
for release builds and Assert for debug builds

Commit:
Check that we don't pass NULL pointer

Commit:
Check that scalarVal is not null

pushJsonbValueScalar dereferences scalarVal without checking that it's not null.
This patch adds an assert

RekGRpth commented on src/backend/utils/adt/jsonb_util.c:
```diff
@@ -585,7 +585,7 @@ pushJsonbValueScalar(JsonbParseState **pstate, JsonbIteratorToken seq,
 			break;
 		case WJB_VALUE:
-			Assert(IsAJsonbScalar(scalarVal));
+			Assert(scalarVal && IsAJsonbScalar(scalarVal));
```
on release build scalarVal is dereferenced in function appendValue

dnskvlnk replied:
Yes, it's better to move the assert there. But what do you think about leaving this as is? IsAJsonbScalar also dereferences scalarVal

dnskvlnk replied:
Removed this line and added Insist to appendValue

Commit:
Check that scalarVal is not null

appendValue dereferences scalarVal without checking that it's not null. We can't
guarantee that scalarVal is always valid. This patch adds Insist

RekGRpth commented on src/backend/utils/adt/jsonb_util.c:
```diff
@@ -676,6 +676,7 @@ appendValue(JsonbParseState *pstate, JsonbValue *scalarVal)
 {
 	JsonbValue *object = &pstate->contVal;
 
+	Insist(scalarVal);
```
@andr-sokolov and I decided that Insist should be moved here before assert.
https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/utils/adt/jsonb_util.c#L588-L589

dnskvlnk replied:
Can you explain why? A couple lines below there's one more `IsAJsonbScalar` call, should we duplicate that insist?

RekGRpth replied:
I think so, because the static analyzer does not see assertions and only noticed an error here.

dnskvlnk replied:
What if we change assert to `Assert(scalarVal && IsAJsonbScalar(scalarVal));` and keep Insist at appendValue so that the check will be at the place it should be?

RekGRpth replied:
may be

dnskvlnk replied:
Done

Commit:
Check that scalarVal is not null

appendValue and appendElement dereferences scalarVal without checking that it's
not null. We can't guarantee that scalarVal is always valid. This patch adds Insist

RekGRpth commented on src/backend/utils/adt/jsonb_util.c:
```diff
@@ -581,15 +581,15 @@ pushJsonbValueScalar(JsonbParseState **pstate, JsonbIteratorToken seq,
 		case WJB_KEY:
-			Assert(scalarVal->type == jbvString);
+			Assert(scalarVal && scalarVal->type == jbvString);
 			appendKey(*pstate, scalarVal);
```
Do we also need to add Insist to appendKey?

dnskvlnk replied:
It's similar to appendValue and I assume that the analyzer would complain about it as well

RekGRpth replied:
No, we primarily fix errors, and at the same time make the static analyzer happy.

RekGRpth replied:
I'm still waiting insist in appendKey...

dnskvlnk replied:
Done

RekGRpth commented:
update PR description

Commit:
Check that scalarVal is not null

appendValue and appendKey dereference scalarVal without checking that it's not
null. We can't guarantee that scalarVal is always valid. This patch adds Insist
for release builds and Assert for debug builds

Commit:
Add Insist in appendKey

RekGRpth commented:
update PR title and description according to all changes

dnskvlnk commented:
@RekGRpth they already contain all the changes

RekGRpth commented:
appendElement was changed too... and fix title

dnskvlnk commented:
@RekGRpth title seems fine

Commit:
Check that scalarVal is not null

appendValue, appendElement and appendKey dereference scalarVal without checking
that it's not null. We can't guarantee that scalarVal is always valid. This
patch adds Insist for release builds and Assert for debug builds

dnskvlnk commented:
@RekGRpth changed commit body
